crates/pathfinder/src/ethereum/api.rs

```diff
@@ -0,0 +1,246 @@
+//! Wrapper for the parts of the [`Web3::eth()`](https://docs.rs/web3/latest/web3/api/struct.Eth.html) API that [the ethereum module](super) uses.
+use crate::retry::Retry;
+
+use std::future::Future;
+use std::num::NonZeroU64;
+use std::time::Duration;
+
+use futures::TryFutureExt;
+use tracing::{debug, error, info};
+use web3::{
+    types::{Block, BlockId, Filter, Log, Transaction, TransactionId, H256, U256},
+    Error, Transport, Web3,
+};
+
+/// Error returned by [`Web3EthImpl::logs`].
+#[derive(Debug, thiserror::Error)]
+pub enum LogsError {
+    /// Query exceeded limits (time or result length).
+    #[error("Query limit exceeded.")]
```

Comment:
Just noticed this form of writing while looking at how you solved the `anyhow::Error::new(...).context(...)` for the Other variant. Where do you get this full sentence writing style? At least thiserror docs hint at lowercase no full stop -style, and I think there was some reasoning behind it but didn't go back to watch any talks on this.

Reply:
:facepalm: honestly I didn't remember & haven't looked at `thiserror` guidelines. Will fix it ofc :+1: .